autoload/cmake/fileapi.vim

```diff
@@ -0,0 +1,192 @@
+" ==============================================================================
+" Location:    autoload/cmake/fileapi.vim
+" Description: Functions for interfacing with the cmake-file-api(7)
+" ==============================================================================
+
+let s:fileapi = {}
+let s:fileapi.cmake_version_supported = v:false
+let s:fileapi.last_index_name = 'unset' " won't compare true
+let s:fileapi.response_files = {}
+let s:fileapi.build_targets = []
+
+let s:client_name = 'vim-cmake'
+let s:api_path = ['.cmake', 'api', 'v1']
+let s:query_path = s:api_path + ['query', 'client-vim-cmake', 'query.json']
+let s:reply_path = s:api_path + ['reply']
+
+let s:query_version = 1
+let s:query = {
+            \     'requests': [
+            \         { 'kind': 'codemodel' , 'version': 2 },
+            \     ],
+            \     'client': {
+            \         'query': { 'version': s:query_version },
+            \     },
+            \ }
+
+let s:logger = cmake#logger#Get()
+let s:system = cmake#system#Get()
+
+""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+" Private functions
+""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+" Find the current index file
+"
+function! s:FindIndexFile(build_dir) abort
+    let l:glob = s:system.Path([a:build_dir] + s:reply_path + ['index'], v:false) . '-*.json'
+    let l:indices = glob(l:glob, v:true, v:true)
+    if len(l:indices) == 0
+        return ''
+    else
+        return l:indices[-1]
+    endif
+endfunction
+
+" Parse api response index
+"
+" Returns:
+"     Number
+"         0: Failed
+"         1: Updated
+"         2: No Update
+"
+function! s:ParseIndex(build_dir) abort
+    let l:index_path = s:FindIndexFile(a:build_dir)
+    if l:index_path ==# ''
+        " TODO: only show this with an existing build tree without api files
+        " and when cmake version new enough
+        let l:warning =
+                    \ 'fileapi: Response from cmake-file-api(7) missing.'
+                    \ . ' Target completion will not work.'
+                    \ . ' Run :CMakeGenerate'
```

Comment:
I realized this warning gets echoed whatever command you run when there isn't a build system yet. Even if you just run `:CMakeGenerate` without a build system being there yet. Which is not what we want I guess.

Ideally, this warning gets echoed only if the user tries to tab-complete targets before generating a build system, i.e. only when this function is triggered as a consequence of invoking `cmake#GetBuildTargets()` (from autoload/cmake.vim). We should not issue this warning if `s:fileapi.Parse()` is invoked from other code paths.

Would you agree? To circumvent this problem, we could only raise an exception from this function, and silently ignore such exception everywhere except for in `cmake#GetBuildTargets()`, where we do echo the warning. Or something else that has the same effect.

Sorry I'm raising this just now, I hadn't tested this properly.